src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -0,0 +1,320 @@
+
+instance walk.inhabited (v : V) : inhabited (G.walk v v) := ⟨by refl⟩
+
+namespace walk
+variables {G}
+
+/-- The length of a walk is the number of edges along it. -/
+def length : Π {u v : V}, G.walk u v → ℕ
+| _ _ nil := 0
+| _ _ (cons _ q) := q.length.succ
+
+/-- The concatenation of two compatible walks. -/
+@[trans]
+def append : Π {u v w : V}, G.walk u v → G.walk v w → G.walk u w
+| _ _ _ nil q := q
+| _ _ _ (cons h p) q := cons h (p.append q)
+
+/-- The concatenation of the reverse of the first walk with the second walk. -/
+protected def reverse_aux : Π {u v w : V}, G.walk u v → G.walk u w → G.walk v w
+| _ _ _ nil q := q
+| _ _ _ (cons h p) q := reverse_aux p (cons (G.symm h) q)
+
+/-- The walk in reverse. -/
+@[symm]
+def reverse {u v : V} (w : G.walk u v) : G.walk v u := w.reverse_aux nil
+
+/-- Get the `n`th vertex from a walk, where `n` is generally expected to be
+between `0` and `p.length`, inclusive.
+If `n` is greater than or equal to `p.length`, the result is the path's endpoint. -/
+def get_vert : Π {u v : V} (p : G.walk u v) (n : ℕ), V
+| u v nil _ := u
+| u v (cons _ _) 0 := u
+| u v (cons _ q) (n+1) := q.get_vert n
+
+@[simp] lemma cons_append {u v w x : V} (h : G.adj u v) (p : G.walk v w) (q : G.walk w x) :
+  (cons h p).append q = cons h (p.append q) := rfl
+
+@[simp] lemma cons_nil_append {u v w : V} (h : G.adj u v) (p : G.walk v w) :
+  (cons h nil).append p = cons h p := rfl
+
+@[simp] lemma append_nil : Π {u v : V} (p : G.walk u v), p.append nil = p
+| _ _ nil := rfl
+| _ _ (cons h p) := by rw [cons_append, append_nil]
+
+@[simp] lemma nil_append {u v : V} (p : G.walk u v) : nil.append p = p := rfl
+
+lemma append_assoc : Π {u v w x : V} (p : G.walk u v) (q : G.walk v w) (r : G.walk w x),
+  p.append (q.append r) = (p.append q).append r
+| _ _ _ _ nil _ _ := rfl
+| _ _ _ _ (cons h p') q r := by { dsimp only [append], rw append_assoc, }
+
+@[simp] lemma nil_reverse {u : V} : (nil : G.walk u u).reverse = nil := rfl
+
+lemma singleton_reverse {u v : V} (h : G.adj u v) :
+  (cons h nil).reverse = cons (G.symm h) nil := rfl
+
+@[simp]
+protected lemma reverse_aux_eq_reverse_append {u v w : V} (p : G.walk u v) (q : G.walk u w) :
+  p.reverse_aux q = p.reverse.append q :=
+begin
+  induction p generalizing q w,
+  { refl },
+  { dunfold walk.reverse_aux walk.reverse,
+    rw [p_ih, p_ih, ←append_assoc],
+    refl, }
+end
+
+@[simp] lemma reverse_append {u v w : V} (p : G.walk u v) (q : G.walk v w) :
+  (p.append q).reverse = q.reverse.append p.reverse :=
+begin
+  induction p generalizing q w,
+  { simp },
+  { dsimp only [cons_append, reverse, walk.reverse_aux],
+    simp only [p_ih, walk.reverse_aux_eq_reverse_append, append_nil],
+    rw append_assoc, }
+end
+
+@[simp] lemma reverse_cons {u v w : V} (h : G.adj u v) (p : G.walk v w) :
+  (cons h p).reverse = p.reverse.append (cons (G.symm h) nil) :=
+begin
+  dsimp [reverse, walk.reverse_aux],
+  simp only [walk.reverse_aux_eq_reverse_append, append_nil],
+end
```

Comment:
```suggestion
@[simp] lemma cons_reverse_aux {u v w x : V} (p : G.walk u v) (q : G.walk w x) (h : G.adj w u) :
  (cons h p).reverse_aux q = p.reverse_aux (cons (G.symm h) q) := rfl

@[simp] protected lemma reverse_aux_eq_reverse_append :
  ∀ {u v w} (p : G.walk u v) (q : G.walk u w), p.reverse_aux q = p.reverse.append q
| _ _ _ nil _ := rfl
| u v w (cons h p) q := by rw [reverse, cons_reverse_aux, cons_reverse_aux,
  reverse_aux_eq_reverse_append, reverse_aux_eq_reverse_append, ←append_assoc, cons_nil_append]

@[simp] lemma reverse_cons {u v w : V} (h : G.adj u v) (p : G.walk v w) :
  (cons h p).reverse = p.reverse.append (cons (G.symm h) nil) :=
by rw [reverse, cons_reverse_aux, walk.reverse_aux_eq_reverse_append]

@[simp] lemma reverse_append : ∀ {u v w : V} (p : G.walk u v) (q : G.walk v w),
  (p.append q).reverse = q.reverse.append p.reverse
| _ _ _ nil _ := by simp
| _ _ _ (cons h p) q := by rw [cons_append, reverse, cons_reverse_aux,
    walk.reverse_aux_eq_reverse_append, reverse_append, ←append_assoc, reverse_cons]
```
These three can be golfed if you add an extra lemma and re-order

Reply:
Oops, I was coincidentally golfing at the same time. I merged our work -- thanks